src/javaRosa.js

```diff
@@ -384,123 +282,137 @@ define([
+                        // this should never happen
+                        window.console.log(
+                            "ignoring ItextItem without a key: " + item.id);
+                        return;
+                    } else if (item && !seen.hasOwnProperty(item.key)) {
+                        seen[item.key] = true;
+                        visit(item, mug, property);
+                    }
+                });
             }
+            processChildren();
+        });
+    }
 
-            var mt = node.getValue();
-            if(!mt) {
-                throw 'Node in tree without value?!?!';
-            }
-        
-            var thingsToGet = [
-                'labelItextID',
-                'hintItextID', 
-                'helpItextID',
-                'constraintMsgItextID'
-            ]; 
-        
-            var val;            
-            for (var i = 0; i < thingsToGet.length; i++) {
-                try {
-                    val = mt.p[thingsToGet[i]];
-                    if (val && !val.isEmpty() && ret.indexOf(val) === -1) {
-                       // it was there and not present so add it to the list
-                       ret.push(val);
-                    } 
-                } catch (err) {
-                    // probably just wasn't in the mug
+    /**
+     * Walks the tree and grabs Itext items from mugs
+     *
+     * This updates the ID of each returned Itext item according to it's
+     * autoId property. IDs of items with autoId turned off will not be
+     * modified unless the ID is blank or it conflicts with another item.
+     * NOTE because this mutates itext IDs it could cause subtle side
+     * effects if anything depends on Itext IDs not changing at random
+     * times such as save, copy, paste, export translations, etc.
+     *
+     * @param form - the vellum instance's Form object.
+     * @param empty - if true, return empty items as well. Otherwise omit them.
+     * @returns - a list of Itext items.
+     */
+    function getItextItemsFromMugs(form, empty) {
+        var ret = [],
+            byId = {},
+            props = _.object(_.map(ITEXT_PROPERTIES, function (thing) {
+                return [thing, thing.replace("Itext", "")];
+            }));
+
+        forEachItextItem(form, function (item, mug, property) {
+            var itemIsEmpty = item.isEmpty();
+            if (!itemIsEmpty || empty) {
+                var id = item.autoId || !item.id ?
+                         getDefaultItextId(mug, props[property]) : item.id,
+                    origId = id,
+                    count = 2;
+                if (itemIsEmpty && byId.hasOwnProperty(id)) {
+                    // ignore empty item with duplicate ID
+                    return;
                 }
-            }    
-        };
-        
-        form.tree.treeMap(appendItemsIfPresent);
+                while (byId.hasOwnProperty(id)) {
+                    id = origId + count;
+                }
```

Comment:
This used to be done by [`deduplicateIds`](https://github.com/dimagi/Vellum/pull/360/files#diff-e5b3eca8407b45b132e0147b8392192cL267)